### ISSUE TITLE
Mat/Pat - CRM ticket field mapping

### ISSUE THREAD

**Issue by fazil-ey** [bcgov/digital-journeys#1224]
The following CRM ticket fields needs to be mapped as follows - 

- [x] Subject should be the name of the form in digital journeys - "Maternity and parental leave form for "Applicant name" 
- [x] Primary contact should be supervisor
- [ ] ~~Secondary reference should be the applicant. Use applicants employee number~~ (moved to #1318)
- [x] Priority due date should be the leave start date 
    - [x] Mat leave start date if taking mat leave
    - [ ] Parental leave start date if taking only parental leave
- [x]  Category should be - TES>Tier 2>T2 HR (Transactional)>Leave & Time Off>Maternity/Parental Leave
- [x]  Assigned - TES HR Admin Technicians

**Comment by Stella-Archer** [comment]
@Iman could you look at this and give us and estimate?

**Comment by iman-jamali-fw** [comment]
@Stella-Archer 
Could you please confirm that the assigned field in the CRM ticket should be "TES HR Admin Technicians"? Thank you

**Comment by iman-jamali-fw** [comment]
@fazil-ey 
For the AC: "**(if available) Secondary reference should be the applicant. Use applicants alternate email id and phone number.**"

Learned from #1277, non-IDIR users lack CRM contact IDs (necessary for ticket contacts). However, Telus can manually add contact IDs for non-IDIR users upon request.

There's a potential for automation: creating a CRM contact for the non-IDIR employee email, then using it as a secondary reference for the CRM ticket. This needs further exploration.

If you prefer the employee's IDIR contact, it can also be added as a secondary reference.

**Comment by fazil-ey** [comment]
i think we can leave the secondary reference as employee information is already available in the thread

**Comment by fazil-ey** [comment]
New request
@iman-jamali-fw can we put the employee number for the applicant as the secondary reference?

**Comment by iman-jamali-fw** [comment]
@fazil-ey 
Yes, we can. We can look up a  user's employId on CRM to find their contact record and assign that contact as a secondary reference. However, that contact record is based on the employee IDIR and the email notification will be sent to the email associated with their IDIR. If the goal is to send a notification to the employee's alternative email address this won't work.

Another point, look up the user's on CRM based on IDIR or empoyeeId will return the same contact record.

**Comment by iman-jamali-fw** [comment]
@fazil-ey 
OK, the employee was added as secondary contact (called `otherContacts` internally in CRM) based on their employeeId.

FYI, I increased the estimated effort to 5.

**Comment by iman-jamali-fw** [comment]
@fazil-ey 
Ready to be tested on TEST. There is a requirement that the employee should have employeeId when initially submitting the Mat/Pat form. So a PSA employee needs to submit the form, we as a contractor cannot at this point (fails on manager validation if submitted by some who doesn't have empolyeeId like you, me, and Bhumin) unless we make it more flexible in the future.

**Comment by iman-jamali-fw** [comment]
@fazil-ey 
`Priority due date` was added to the CRM ticket according to the AC. Ready to be tested on TEST

**Comment by iman-jamali-fw** [comment]
@ayushdamani 
Please follow the testing instructions below on TEST environment:

1. Submit a Mat/Pat form.
2. Ask the manager you entered in the form, to validate that submission. Once the submission is validate, a CRM ticket should be created in the background. You need to follow the next steps to find that CRM ticket id and verify its details against the Acceptance criteria in this ticket.
3. To find the `id` of the CRM ticket you need to go to the "Submitted Forms" tab and find your Mat/Pat submission in the list. It should have "Validated" status. Click Edit to open that submission. Open your browser's network tab and refresh the page, find the form data in one of the network calls, under data, property name `crmId`. Take note that id.
4. Open CRM UI (ask Andrew if you don't have access to it) and look for that CRM ticket by the `id` you found in the last step. Verify the CRM ticket details against this ticket's Acceptance criteria.
5. For the "Priority due date" you can send two Mat/Pat submissions, one for a case of Maternity leave and check the Mat leave start date should be in the "priority due date" filed in the CRM ticket. Send another one that is Parental Leave only, This time, check your CRM ticket to have the Parental Leave start date set in the `Priority due date` field.


Good luck with your testing :)

**Comment by ayushdamani** [comment]
Thank you, Iman!
For some reason, I was not able to validate my own submission. I have summited another form with Chris as my supervisor. Have requested him to validate the submission.

**Comment by ayushdamani** [comment]
Thank you, Iman!
For some reason, I was not able to validate my own submission. I have summited another form with Chris as my supervisor. Have requested him to validate the submission.

**Comment by ayushdamani** [comment]
Tried to test with Iman, but couldn't submit the form. Iman will contact Andrew regarding an error - 'employee contact id'

**Comment by iman-jamali-fw** [comment]
@ayushdamani 
I investigated the issue you had on validating the Mat/pat and found an interesting edge case, preventing your CRM ticket generated.

The reason Mat/Pat validation failed for you lies in your CRM contact record. Your CRM contact record, created by Nancy, doesn't have all your employee information, including your `employeeId`. When you first submitted the Mat/Pat as an employee, your `employeeId` was added to your submission as a hidden field. When you came back to validate the same submission as the manager, our CRM service in the background assumed that you should have the same `employeeId` in the CRM and when not find it, it threw an error that said `employeeContactId is null`. Therefore, any Mat/Pat submission that is submitted by your account as an employee will fail when the manager (whether it's you or someone else) validates it.

As this is an edge case, I don't think we need to add additional logic in the backend. Your CRM contact record is OK on the Production.

To fix this, please get in touch with Nancy and ask her to add your `employeeId` to your CRM contact record. That should fix the issue you're experiencing. 

To continue with the testing, while your issue is being fixed by Nancy, you can ask someone else on the team to submit a Mat/Pat submission and put you as manager.

**Comment by MeghanStothers** [comment]
Ayush send me the email to Nancy CRM - Meghan  to reach out to Nancy

**Comment by chriscaldwell-psa** [comment]
Submitted a Mat/Pat Form to Ayush as my manager as indicated in the instructions above by Iman. =

**Comment by iman-jamali-fw** [comment]
@ayushdamani @chriscaldwell-psa 
Did you have a chance to test this out?

**Comment by ayushdamani** [comment]
Chris and I tried to test it out. We are not able to validate the submission. Tried two different submissions, one where I am the supervisor and the other where is the supervisor.

**Comment by MeghanStothers** [comment]
@Ayush - can you please repeat and document the steps you took so we can understand your testing method and uncover the issue? Thanks

**Comment by ayushdamani** [comment]
For sure, Meghan. Here are the details -
1. Submitted a form with Chris as my supervisor.
2. Chris tried to validate the submission, but failed.

Error noticed: When Chris hits the "validate submission" button, the page loops back to the "before you begin" (first page), and thus does not validate the submission.

We also tried -
1. Chris submitted a form with me as a supervisor.
2. I tried to validate the submission, but failed.

Same error.

**Comment by ayushdamani** [comment]
The team at check-in and Iman tested this out together.

We were able to test the ticket, as per Iman's instructions. The data in the CRM UI matches the acceptance criteria in the ticket ✅

[Private Zenhub Image](https://api.zenhub.com/attachedFiles/eyJfcmFpbHMiOnsibWVzc2FnZSI6IkJBaHBBdGR5IiwiZXhwIjpudWxsLCJwdXIiOiJibG9iX2lkIn19--18faef6c662ee80cf21071049832a58c31b18d3c/MicrosoftTeams-image%20%281%29.png)

**Comment by fazil-ey** [comment]
Will confirm if employee id is in secondary reference as its not in the screenshot

**Comment by fazil-ey** [comment]
Will confirm if employee id is in secondary reference as its not in the screenshot

**Comment by ayushdamani** [comment]
To open CRM UI:
- Go to this link https://psa--rnt2.frontlinesvc.com/cgi-bin/psa.cfg/php/admin/launch.php and install Oracle Cloud Service.
- Login using credentials
username: crmapi
password: Password2
- Use the ticket number: 230906-000002
- Compare acceptance criteria to the details in the CRM UI

**Comment by fazil-ey** [comment]
@iman-jamali-fw  while testing we noticed there was nothing in the secondary reference.

**Comment by MeghanStothers** [comment]
[Uploading... image.png]

**Comment by iman-jamali-fw** [comment]
Sorry, I confused "Secondary Reference" with "Secondary Contact". I reached out to Nancy to figure out which field in the CRM API best maps to "Secondary Reference".

**Comment by iman-jamali-fw** [comment]
Madhu Dorai from CRM team get back. What I learned so far is that "secondary reference" is a custom object stored in a separate table with a relationship to the CRM indecent (ticket) table. Adding that field is a little different from the fields we've added so far (these were all part of the ticket table).

I'm trying to get hold of Madhu Dorai to learn how to leverage API to add secondary reference. Meanwhile, I'd suggest splitting this work into another ticket. With the CRM ticket reapproval 1107 on the way, we may want this field to be added and updated (on mat/pat reapproval).

**Comment by iman-jamali-fw** [comment]
@fazil-ey @ayushdamani 
Before splitting the "secondary reference" into a separate ticket, can you please verify the **_Parental leave start date if taking only parental leave_** AC? I see that one is not checked, was there any issue or you haven't had the chance to review it?

**Comment by iman-jamali-fw** [comment]
Move AC **_Secondary reference should be the applicant. Use applicants employee number_** to a separate ticket #1318

**Comment by ayushdamani** [comment]
Since I am not able to validate my own submission, I will request the team at check-in to run another test.
Thank you, Iman!

**Comment by ayushdamani** [comment]
1. Tested with the team at check-in.
2. Fazil submitted a form with Meghan as the supervisor.
3. Meghan validated the submission.
4. We were able to find the crmid.
5. Stella checked the crmid in CRM.

[Private Zenhub Image](https://api.zenhub.com/attachedFiles/eyJfcmFpbHMiOnsibWVzc2FnZSI6IkJBaHBBbytYIiwiZXhwIjpudWxsLCJwdXIiOiJibG9iX2lkIn19--6a064a9d1db2dd161f440c4e529609274e984659/Screenshot%202023-09-13%20075546.png)

As advised by Fazil, closing the ticket.
